src/main/java/com/linecorp/armeria/client/SessionProtocolNegotiationCache.java

```diff
@@ -0,0 +1,196 @@
+    private static final Map<String, CacheEntry> cache = new LruMap<String, CacheEntry>(65536) {
+        private static final long serialVersionUID = -2506868886873712772L;
+
+        @Override
+        protected boolean removeEldestEntry(Entry<String, CacheEntry> eldest) {
+            final boolean remove = super.removeEldestEntry(eldest);
+            if (remove) {
+                logger.debug("Evicted: '{}' does not support ", eldest.getKey(), eldest.getValue());
+            }
+
+            return remove;
+        }
+    };
+
+    /**
+     * Returns {@code true} if the specified {@code remoteAddress} is known to have no support for
+     * the specified {@link SessionProtocol}.
+     */
+    public static boolean isUnsupported(SocketAddress remoteAddress, SessionProtocol protocol) {
+        final String key = key(remoteAddress);
+        final CacheEntry e;
+        final long stamp = lock.readLock();
+        try {
+            e = cache.get(key);
+        } finally {
+            lock.unlockRead(stamp);
+        }
+
+        if (e == null) {
+            // Can't tell if it's unsupported
+            return false;
+        }
+
+        return e.isUnsupported(protocol);
+    }
+
+    /**
+     * Updates the cache with the information that the specified {@code remoteAddress} does not support
+     * the specified {@link SessionProtocol}.
+     */
+    public static void setUnsupported(SocketAddress remoteAddress, SessionProtocol protocol) {
+        final String key = key(remoteAddress);
+        final CacheEntry e = getOrCreate(key);
+
+        if (e.addUnsupported(protocol)) {
+            logger.debug("Updated: '{}' does not support {}", key, e);
+        }
+    }
+
+    /**
+     * Clears the cache.
+     */
+    public static void clear() {
+        int size;
+        long stamp = lock.readLock();
+        try {
+            size = cache.size();
+            if (size != 0) {
+                final long writeStamp = lock.tryConvertToWriteLock(stamp);
+                if (writeStamp == 0L) {
+                    lock.unlockRead(stamp);
+                    stamp = lock.writeLock();
+                } else {
+                    stamp = writeStamp;
+                }
+                size = cache.size();
+                cache.clear();
+            }
+        } finally {
+            lock.unlock(stamp);
+        }
+
+        if (size != 0 && logger.isDebugEnabled()) {
+            if (size != 1) {
+                logger.debug("Cleared: {} entries", size);
+            } else {
+                logger.debug("Cleared: 1 entry");
+            }
+        }
+    }
+
+    private static CacheEntry getOrCreate(String key) {
+        long stamp = lock.readLock();
+        try {
+            final CacheEntry entry = cache.get(key);
+            if (entry != null) {
+                return entry;
+            }
+
+            final long writeStamp = lock.tryConvertToWriteLock(stamp);
+            if (writeStamp == 0L) {
```

Comment:
Hmm I wonder what use case there is for the SDK not doing this for us... Anyways, can you define a private method convertToWriteLock to share the code?